Published asset deletion protection design doc (#2586)
dandi/dandi-archive: doc/design/s3-object-lock-legal-holds.md

# S3 Object Lock Legal Holds

## Terminology
The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt).

## Why make published blobs immutable?

The DANDI Archive's core value proposition is hosting reliable, citable scientific data. Once a dandiset is published and assigned a DOI, the data it references MUST remain immutable to preserve the integrity of scientific citations and reproducibility. Users who cite a published dandiset expect that the data will remain exactly as it was at the time of publication.

> Currently, while our application logic prevents modification of published assets, there is no infrastructure-level enforcement.  Adding S3-level immutability provides stronger data safety.

Reviewer: I feel it is jumping over the different levels (S3: blobs, zarr prefixes; and SQL DB: assets associated with those) without due description. May be smth like

 ```suggestion
Dandiset is a collection of assets which are associated many-to-one to content on S3 of two data types: blobs and zarrs.
A blob is a representation of a single key ("file") under `blobs/` prefix, while zarr is a prefix ("folder") under `zarr/` prefix).
`blobs/` store is already immutable, as existing keys MUST NOT mutate, while zarrs are allowed to be updated under `zarr/`.
Only a single blob or a single zarr is associated with a single asset.
Whenever zarr assets remain mutable and thus cannot be published ATM, published blob assets MUST remain immutable and our application logic already prevents modification of published assets.
But there is no infrastructure-level enforcement to guarantee immutability of published blobs on S3.  Adding S3-level immutability provides stronger data safety.
```

## Requirements

- Published asset blobs MUST be immutable at the S3 infrastructure level
- The immutability mechanism MUST not interfere with the existing trailing delete functionality (S3 Bucket Versioning)
- The solution MUST be compatible with future garbage collection features
- The mechanism SHOULD be efficient when applied to large numbers of blobs during publication
- There MUST be some sort of "backdoor" that allows deletion of published blobs if necessary, in case of extreme circumstances that warrant deletion of a published asset and/or dandiset

## Proposed Solution

> AWS S3 Object Lock with legal holds provides exactly the protection we need. A legal hold is a flag that can be applied to an S3 object version to prevent deletion or modification, regardless of any retention policies or IAM permissions. Unlike retention modes, legal holds have no expiration date and remain in effect until explicitly removed via the `s3:PutObjectLegalHold` S3 API. See the [documentation for S3 legal holds](https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-lock.html#object-lock-legal-holds) for more details.

Reviewer: Although those are indeed two neighboring subsections, new reader would not know of such proximities reference would be handy IMHO, especially since "retention mode" keeps coming up.

```suggestion
AWS S3 Object Lock with legal holds provides exactly the protection we need. A legal hold is a flag that can be applied to an S3 object version to prevent deletion or modification, regardless of any retention policies or IAM permissions. Unlike [retention modes](https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-lock.html#object-lock-retention-modes), legal holds have no expiration date and remain in effect until explicitly removed via the `s3:PutObjectLegalHold` S3 API. See the [documentation for S3 legal holds](https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-lock.html#object-lock-legal-holds) for more details.
```

### S3 Object Lock

S3 Object Lock is a feature that prevents object versions from being deleted or overwritten for a specified period or indefinitely. It operates in three ways:

1. **Governance mode**: Prevents most users from deleting object versions, but users with special permissions can override
2. **Compliance mode**: No user can delete the object version until the retention period expires, including the root account
3. **Legal hold**: A separate flag independent of retention mode that prevents deletion until removed

> Goverence and compliance modes operate at the bucket level, while legal holds operate at the object level.

Reviewer: ```suggestion
Goverence and compliance modes operate at the bucket level, while legal holds operate at the object version level.
```

IMHO an important feature (if I got it right) since will allow later to protect Zarr versions. (added below as an item)

For our use case, **legal holds** are the appropriate choice because:
- Only published assets need protection from deletion, other objects in the bucket MUST remain deletable
- Published assets MUST remain immutable permanently (no expiration)
- There MUST be the ability to remove the hold if absolutely necessary (per the "backdoor" requirement)

> - Legal holds are independent of retention policies and work alongside bucket versioning

Reviewer: ```suggestion
- Legal holds are independent of retention policies and work alongside bucket versioning
- Legal holds apply to an object version thus MAY later be adopted for protection of published zarr assets
```

## Interaction with Garbage Collection

Legal holds MUST NOT interfere with garbage collection of unpublished assets, as only published asset blobs SHALL have legal holds applied. The garbage collection processes for orphaned uploads, unreferenced AssetBlobs, and unpublished assets MUST continue to work as designed.
Any violation of this, i.e. any interference, as e.g. attempt to remove a blob which is under Legal hold MUST be considered to be a bug as MUST NOT happen under current design, and thus will not be silently ignored. 

### Potential Interference Scenarios (and Why They Don't Occur)

> The following scenarios could theoretically interfere with garbage collection, but do NOT occur with this design:

Reviewer: ```suggestion
The following scenarios MAY theoretically interfere with garbage collection, but SHOULD NOT occur with this design:
```

or should even be made 

```suggestion
The following scenarios MAY theoretically interfere with garbage collection, but MUST NOT occur with this design:
```

thus warranting ERROR reporting?

> 1. **Overly broad legal holds**: If legal holds were accidentally applied to unpublished blobs, those blobs could not be deleted by garbage collection processes until the holds were manually removed.

Reviewer: but I am thinking about deleting a published dandiset overall (e.g. per author's or regulatory organ instruction).  We would need to implement a process to review if each blob has any other "publication" or even just a "draft" version and potentially be able to decide on blob-by-blob case on what to do.

I think it is worth adding some statement here that we are not aiming to design/implement that logic here but that it MAY be designed/implemented later

```suggestion
1. **Overly broad legal holds**: If legal holds were accidentally applied to unpublished blobs, those blobs could not be deleted by garbage collection processes until the holds were manually removed.
     - Under force-major circumstances requiring deletion of a published dandiset, and thus its assets (all or some), legal holds might remain attached to assets associated with published or not blobs. Analysis and design of handling of such cases MAY be implemented later, but overall the mechanism of legal holds seems to be appropriate and will not become an obstacle.
```

# S3 Object Lock Legal Holds

## Why make published blobs immutable?

The DANDI Archive's core value proposition is hosting reliable, citable scientific data. Once a dandiset is published and assigned a DOI, the data it references must remain immutable to preserve the integrity of scientific citations and reproducibility. Users who cite a published dandiset expect that the data will remain exactly as it was at the time of publication.

Currently, while our application logic prevents modification of published assets, there is no infrastructure-level enforcement.  Adding S3-level immutability provides stronger data safety.

## Requirements

- Published asset blobs must be immutable at the S3 infrastructure level

> - The immutability mechanism should not interfere with the existing trailing delete functionality (S3 Bucket Versioning)

Reviewer: FWIW, might be worth following [RFC2119](https://www.ietf.org/rfc/rfc2119.txt) overall in design docs and CAPITALIZE those MUST, SHOULD, MAY..., WDYT?

- The solution should be compatible with future garbage collection features
- The mechanism should be efficient when applied to large numbers of blobs during publication
- There should be some sort of "backdoor" that allows deletion of published blobs if necessary, in case of extreme circumstances that warrant deletion of a published asset and/or dandiset

## Proposed Solution

AWS S3 Object Lock with legal holds provides exactly the protection we need. A legal hold is a flag that can be applied to an S3 object to prevent deletion or modification, regardless of any retention policies or IAM permissions. Unlike retention modes, legal holds have no expiration date and remain in effect until explicitly removed.

### S3 Object Lock

> S3 Object Lock is a feature that prevents objects from being deleted or overwritten for a specified period or indefinitely. It operates in three ways:

Reviewer 2: Why would this affect Zarrs? Aren't we planning to use Dandidav to implement versioned Zarrs (rather than S3 object versioning)?

1. **Governance mode**: Prevents most users from deleting objects, but users with special permissions can override
2. **Compliance mode**: No user can delete the object until the retention period expires, including the root account
3. **Legal hold**: A separate flag independent of retention mode that prevents deletion until removed

Goverence and compliance modes operate at the bucket level, while legal holds operate at the object level.

For our use case, **legal holds** are the appropriate choice because:
- We only want to prevent deletion of published assets, other objects in the bucket should remain deletable
- Published assets should remain immutable permanently (no expiration)
- We still want the ability to remove the hold if absolutely necessary
- Legal holds are independent of retention policies and work alongside bucket versioning

## Interaction with Garbage Collection

> Legal holds do not interfere with garbage collection of unpublished assets, as only published asset blobs will have legal holds applied. The garbage collection processes for orphaned uploads, unreferenced AssetBlobs, and unpublished assets will continue to work as designed.

Reviewer 2: The idea is that, by design, only published assets will have the legal hold applied to them. GC only affects other kinds of objects (specifically, our GC logic will not attempt to reclaim objects that have a legal hold on them).

The point here is there will be no "interference". If we made a mistake in the GC logic, we'll know because GC will fail on those objects.

> Legal holds do not interfere with garbage collection of unpublished assets, as only published asset blobs will have legal holds applied. The garbage collection processes for orphaned uploads, unreferenced AssetBlobs, and unpublished assets will continue to work as designed.

Author: https://github.com/dandi/dandi-archive/pull/2586/commits/40be88437bd7675fdad0a5b37e5ffe4f825b3594
